035-cluster-wide-volume-usage-quota-managemnet.md

```diff
@@ -0,0 +1,129 @@
+## Proposal
+
+### Leverage kafka to distribute volume usage metrics throughout the cluster.
+
+By publishing per volume usage metrics to a compacted topic keyed by the broker ID each instance of the plugin will be able
+to start and quickly determine the state of the cluster and consistently apply throttling regardless of where the client
+is connected.
+
+[KIP-257](https://cwiki.apache.org/confluence/display/KAFKA/KIP-257+-+Configurable+Quota+Management) Defines the quota
+API and defines quotas in terms of how much of a delay to apply to a given produce request. This leads to defining the
+possible states of the plugin as:
+`OPEN`, `THROTTLE` and `PAUSE` to accurately reflect the effects of transitioning between states.
+
+#### Limit Types
+
+1. Consumed space limit: Triggers if the consumed space of a volume breaches the value. Candidate for deprecation.
+2. Minimum Free Bytes: Triggers if the amount of free space on a volume drops below the configured level. Particularly
+   useful for hard limits as an absolute minimum of free space.
+3. Minimum Free Percentage. Triggers if the amount of free space on the volume drops below the configured proportion of
+   the Volumes total capacity.
+
+#### The quota plugin data publisher should:
+
+- Collect local volume usage
+- Publish usage to an internal compacted topic keyed by broker ID.
+- Collect and publish usage on start up and periodically after that.
+- Connect as an internal service using TLS and client certificates to connect to the Control Plane listener.
+- Publish metrics showing its connection status and a counter of the number of publication errors - to allow the
+  detection of connectivity issues.
+- Publish Volume Information with the following [schema](#Message-schema)
+
+#### The quota plugin consumer should:
+
+- Should apply the largest delay caused by each disk which breaches the soft limit.
+    - The amount of throttling should be proportional to how much of the difference between the soft and hard limit
+      remains for a given Volume. Applying more throttle the closer to the hard limit things are.
+- Should apply a `PAUSE` of publication if *any* Volume in the cluster breaches the hard limit
+- Connect as an internal service using TLS and client certificates to connect to the Control Plane listener.
+- Use the Kafka Admin client to discover the list of currently active brokers
+- On startup and periodically thereafter read the Volume usage stats from the topic
+- Ensure that the quota plugin Volume updates are not throttled
+- Should take a configured action when it can not determine the state of all known brokers.
+    - The action should be one of `PAUSE` or `OPEN`.
+    - Defaulting to `PAUSE` as the fail-safe option.
+- Publish a metric showing when it is applying throttling due to a lack of data.
+- Apply a freshness check to the volume usage it reads and ignore stale state [3]. The threshold for staleness
+  should be a configuration parameter
+
+#### Message schema
```

Comment:
Is it going to be a binary message or using JSON, XML or something else?

Reply:
covered in https://github.com/strimzi/proposals/pull/51#discussion_r833855911